DM-34951: propose new Butler component division of responsibilites

TallJimbo opened the pull request.

TallJimbo committed:
Switch abstract to linebreaks after sentences.

TallJimbo committed:
Add first draft of section on Registry/Datastore consistency.

TallJimbo committed:
Add stubs for public interfaces and QuantumDirectory sections.

TallJimbo committed:
Add DMTN-205 to bib.

ktlim commented on index.rst:
```diff
@@ -13,17 +13,249 @@
+
+2. Call ``Datastore.get`` with both the resolved ``DatasetRef`` and the bundle of records, returning the result to the caller.
+
+Because this is a write operation consistency in the presence of failures is not a concern, but this still has a major advantage over the current approach for client-server in particular, as it bundles all http server access into a single call, followed by a direct object-store call, reducing latency and again allowing the Datastore to be a regular ``FileDatastore``.
```
I think there's a missing "not" here.  Or `write` needs to be `read`.

ktlim reviewed (approved):
I ❤️  the clear statement of allowable states and the reorganization of transactions.

TallJimbo committed:
Minor edits: fix RFC reference, add missing negation.

timj commented on index.rst:
```diff
@@ -13,17 +13,249 @@
+1. A dataset can be present in either Registry or Datastore without being present in the other.
+
+2. A dataset present in Datastore alone must have a data ID that is valid in the Registry for that data repository (i.e. it uses valid dimension values) and it must not have any Datastore records in the Registry database.
+   This state is expected to be transitory, either intentionally (e.g. during batch execution, before datasets are transferred back), or as a result of failures we cannot rigorously prevent.
```
Or transitory during a put operation if we are moving registry updates to later. Probably should note that datastore should go back to refusing to write a file if a file already exists in that place in the datastore.

TallJimbo replied:
Agreed; I think I now cover both of these points later in the text.

timj commented on index.rst:
```diff
@@ -13,17 +13,249 @@
+
+2. A dataset present in Datastore alone must have a data ID that is valid in the Registry for that data repository (i.e. it uses valid dimension values) and it must not have any Datastore records in the Registry database.
+   This state is expected to be transitory, either intentionally (e.g. during batch execution, before datasets are transferred back), or as a result of failures we cannot rigorously prevent.
+   Datasets in this state as a result of failures or abandoned batch runs are considered undesirable but tolerable, and an approach to minimizing them will be introduced later in :ref:`adding_journal_files`.
```
We likely should write an admin tool that can do an integrity check (although it has to work with millions of files)

TallJimbo committed:
Update Butler.put to include signed URIs and conflict resolution.

TallJimbo commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
```
@timj , I've tried to incorporate signed URIs into my new order of operations (here, and also just below in step 4), and I think this is the only way to do it without introducing another server call to obtain signed URLs before the initial Datastore write.

But this way requires a staging area, which seems like it's a bit wasteful, especially if we want to apply the same approach to bulk operations like `transfer_from` and `import`.  But I'm not actually certain if the staging area you'd envisioned on DMTN-242 was intended to be visible to the server, or if it was added just to help with consistency between Registry and Datastore given the current order of operations and Datastore clobber behavior.

If that's the case - i.e. we don't need a staging area for some other reason - I don't think we need it for this order of operations, and it'd probably be better for me to rip this change out and insert a request for signed URLs before the initial `Datastore.put` instead, so it can use those write to the final location directly.

ktlim replied:
Is the staging area client-side (but with Datastore tables updated?!) or server-side?  A client-side staging area is not terrible, as `resources` writes to local storage anyway before upload, but I'd very much like to avoid a server-side one (which is I thought what DMTN-242 was suggesting), particularly if it uses unsigned URIs (i.e. a dumping ground for any user — or anyone on the Internet?).

timj replied:
I assumed the staging area was on the server since it seemed unlikely that the server would be able to pull files from a staging area on the local laptop. I was using a staging area because I was worried about race conditions with two concurrent puts of the same dataset -- a staging area lets the server Butler implementation move files as part of it's registry transaction and delete them if there was a problem.

If we are deciding that it is possible for registry to get out of sync with datastore temporarily then the staging area can go away.

* Datastore determines how many signed URLs it needs.
* Requests them using datastore server API (maybe passing in the "path in root" part of the file name so the files end up in the right place).
* Server issues signed URLs that are for the final destination (checking that nothing is there already).
* Client serializes to those locations.
* Client returns the appropriate records to Butler layer which passes to registry.

This works much simpler than having to get the butler to hold a transaction.

For races we can probably say that we only issue one signed URL for that specific location and wait for it to be used or expire before we issue another one.

ktlim replied:
That "passing in" part reminded me that we need to make sure that we never trust anything passed in through the c/s Butler API — we must check for access before issuing URLs.  (Well, we can usually trust opaque, long, randomish numbers/strings that we gave out previously.)

timj replied:
Always. This is covered a little in DMTN-242. It is assumed that all server APIs take an auth token and then check collection accessibility.

TallJimbo replied:
I'm totally on board with rewriting this to have the Datastore asking for signed URLs in advance (or something equivalent; maybe Butler does it for Datastore?) instead of having a staging area.

I don't have a useful opinion on where we try to handle conflicts when signed URLs are in play.  But my previous mantra of, "it needs to be solid in terms of consistency and races, but not necessarily fast" for that conflict-handling still applies.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+1. A dataset can be present in either Registry or Datastore without being present in the other.
+
+2. A dataset present in Datastore alone must have a data ID that is valid in the Registry for that data repository (i.e. it uses valid dimension values) and it must not have any Datastore records in the Registry database.
```
We have to be clear that changing the datastore config after a file has been written but before it is stored in registry is going to make the dataset unfindable (template change) or unreadable (formatter change). For batch we understand that the transfer job runs with the same config as that used by the batch jobs.

TallJimbo replied:
To be more defensive about this, we could start copying the relevant bits of datastore configuration into the journal file that points to the QG.

timj replied:
Writing the predicted output records to the graph might work. It might also be that a standard Butler with a graph-backed Registry object would work in the new Butler design.

We could think about a `datastore.put()` that takes optional records and then ignores all the file template and formatter selection configuration but it would till have to know about formatter write parameters (such as data compression settings) so I don't think it's possible to get away from any datastore configuration -- unless you put that in the graph as well and then have that be passed into the Datastore when Butler is created.

TallJimbo replied:
Latest draft does have us putting predicted Datastore records into the QuantumGraph, and the Datastore can't really ignore them (unless it reproduces them instead) since they'd also be used for reads by other Quanta.  That might be enough for a Datastore to catch an inconsistency between its configuration at QG-gen time and its configuration at execution time.  Stuffing Datastore configuration into a QG also doesn't seem too hard, though it might be safer to use that for a consistency check instead of actually allowing to to be used when the main repo's Datastore configuration has changed.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+2. When a dataset is present in both Registry and Datastore, the Registry is fully responsible for storage of Datastore records.
+   It will be the job of the Butler to accept records from Datastore and pass them to Registry on ``put``, and to fetch them from Registry and pass them to Datastore on ``get``.
+   Datasets in this state must always have Datastore records present in the registry, even if the Datastore otherwise has no need for records; this allows a database query to reliably return only datasets that actually exist in a Datastore via a join against the record tables.
```
There needs to be a way for chained datastores to function, which means that the records have to be grouped by datastore name such that a datastore can return something like `dict[str, list[StorageRecords]]` and accept the same (and complain if it is given records for some other datastore and not itself).

`InMemoryDatastore` has no records at all -- the dataset ref ID is the key into an internal cache. Are we really making it return records or are we going to say that `InMemoryDatastore` never returns records and so never exists in registry? We have a use case where you `datastore.put()` to a chain and it gets accepted by in-memory and rejected by fileDatastore so we can't decide to require that in-memory hides inside a chained datastore but fileDatastore will always return a record. We don't really want ephemeral datasets getting explicit entries in registry for in-memory datastore since they disappear as soon as the process dies (even if the registry entry survives for the DatasetRef because it was an intermediate.

A SasquatchDatastore doesn't need to return record information either (it will store the dataset UUID in its system) but I guess you can return a record that is just empty but with the UUID in it if you need it for joins in registry?

TallJimbo replied:
:+1: to grouping the records by datastore.

I think with this plan InMemoryDatastore doesn't return records; if we ever have a use case for running it with a full-butler (not just QBB), then we'll want to be careful with the interfaces that Registry will use to determine how to query for existence-in-Datastore.  We should be able to give it an "iteration engine" relation that makes the query system handle this transparently (e.g. via a join to a temporary table holding the InMemoryDatastore's UUIDs, populated as needed).  But if InMemoryDatastore is limited to QBB we won't need even that.

I do think we'd need SasquashDatastore to provide records, even if the only column is the UUID.  We could put that in a base class or to make it so datastores with trivial records could inherit it, if we anticipate having a few of those.

timj replied:
The base class defaulting to a single column for the record seems fine to me. InMemoryDatastore being allowed to not return records also seems fine (ChainedDatastore could be clever about it).

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+We could use that information with the new ``daf_relation`` classes to easily integrate them with the query system, allowing user queries to not just test for Datastore existence, but query on and report Datastore specific-fields like file size.
+We'd also of course provide a way for users to inspect which such fields are available, since Datastore record fields can change from implementation to implementation.
+
+Datastore methods that add new datasets to the repository would be modified to return a collection of records describing those datasets, again for Butler to pass to Registry.
```
Note that one dataset can return multiple records and for a chained datastore maybe multiple records in multiple datastores.

TallJimbo replied:
I think that's fine, though it might be future source of duplicate results when we join to the record table to check datastore existence.

timj replied:
Yes, you end up having to deal with disassembled composites in registry results again... but this time potentially multiplied by the number of datastores in the chain.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+Datastore methods that add new datasets to the repository would be modified to return a collection of records describing those datasets, again for Butler to pass to Registry.
+Datastore methods that read datasets or interpret the records describing them would be modified to accept those records from Butler (which fetches them from Registry).
+Some Datastore existence-check methods would go away entirely (e.g. ``knows``), as their functionality is subsumed by Registry dataset queries, while others would change their behavior to checking for artifact existence *given* records.
```
Much of this is moving the existing record access methods up from datastore to butler.

TallJimbo replied:
Yes, I think that's exactly right.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
```
Typo: "this is done"

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
```
Datastore always writes to a local temp file first before relocating it to the final location. In POSIX it does that by writing a temp file name into the final directory location so it can do a rename -- if the destination already exists I don't think this can fail so we might not be able to safely do a check for race conditions. For object stores it writes locally if it needs to or can write bytes directly to the remote file store.

TallJimbo replied:
Ah, I hadn't realized `os.rename` couldn't be made to raise in the presence of a clash.  That could indeed be a problem, but I'm hoping there's some recipe for doing some kind of noclobber atomic file creation.

I'm not concerned if we write local temporaries for object stores and then only later have a failure - I see this as being entirely about data repository consistency, as we should be able to avoid conflicts at a higher level (QG generation, explicit retry-handling in pipetask) almost all of the time, so what remain needs to be rock-solid safe but not necessarily efficient.

ktlim replied:
Ah, on the same POSIX filesystem, you can hard link `src` to `dst`, which will fail if `dst` exists.  Then you remove `src`.

TallJimbo replied:
I've updated the text to reflect our more recent Slack conversation about this.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+This has two major advantages over our current ``put`` implementation:
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
```
Is datastore still meant to clean up after itself? If it writes 4 files and somehow fails to write the 5th, should it attempt to delete the previous four?

TallJimbo replied:
I don't think so, but that isn't a strong opinion. If we have some scenario or use case that makes that kind of behavior common, then it might be convenient to clean up immediately, and there's certainly nothing wrong with trying to clean up immediately.  But we can't rely on it to guarantee consistency, and my _suspicion_ is that once we have ways to clean up orphaned Datastore-only files, the automatic-undo logic won't be pulling its weight in terms of code complexity and maintenance.

timj replied:
We might have to reorganize datastore writes so that it writes the N temporary files for disassembled composites and then renames (or uploads) them at the end.

TallJimbo replied:
:+1: to writing all artifacts for a particular dataset before renaming/uploading/hard-linking them to their final location.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
```
As discussed in DMTN-242, do we want an initial ping to registry asking whether this new DatasetRef is already in conflict with registry? Failing early when we know it will fail later seems like a good idea. Obviously in a race the absence of a registry entry does not mean there won't be one when the registry update happens at the end but it will at least potentially save us the bother of doing the slow serializations. This check would be very fast for QBB scenario and for client server will cause a small delay.

TallJimbo replied:
Given that the initial ping is still subject to race conditions, I'm not convinced adding it helps us.  In QBB checking the registry doesn't tell us anything useful, and I don't have a sense for what use case we're optimizing outside of that.  It's still going to be a hard error for users in notebooks, so saving fractions of seconds there isn't going to be noticeable.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
+
+- for a client/server butler, there is no alternation between object-store and http operations, just one and then the other, reducing latency (assuming the data ID has indeed been obtained in advance), and increasing the possiblity that the client Datastore can just be a regular ``FileDatastore``.
```
I'm fairly certain it can't be a FileDatastore but I think it can be 99% FileDatastore with a small interface handling the URLs in a subclass. I don't see any issue with that given that we already know exactly how to switch in different datastores from configuration.

TallJimbo replied:
Yup, after our discussion last Thursday I'm much, much less bothered by having a special Datastore, because I think it's special because it's talking to a special auth-protected storage endpoint, not because it's part of a client/server butler.

In fact, I wonder if "give me a signed URI for this dataset" should be a completely different server from the main butler server, just to establish that separation of concerns further.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+This change will make it so Datastore sees conflicting writes before Registry.
+While these are expected to be rare, and thus conflict resolution does not need to be highly optimized, is important that it does not have the potential to corrupt the repository.
+The easiest way to do this is to configure Datastore storage backends to prohibit writes to locations where files already exist; this is the default behavior for most backends of interest, but we currently explicitly permit silent clobbering to work around problems with automatic retries in BPS.
```
No. We always allow overwrite now. Allowing for no-overwrite can sometime result in obvious races and as I say above I don't think `os.rename` for local POSIX datastores has the ability to prevent rename to an existing file (short of using chmod/umask on every file in the datastore as we write it to make it read-only -- requiring deletes to chmod it back to read/write).

ktlim replied:
Um, I believe rename is purely a directory operation.  It pays no attention to file permissions.

TallJimbo replied:
I think the text agrees with @timj's comment for object stores; I think those "obvious races" go away when we require the Datastore write to succeed before we start (let alone commit) the database transaction.

But as Tim noted in another comment, we may indeed have a problem for POSIX.  I haven't looked hard for solutions there.

And the possibility of new and old butler code writing to the same data repository with Datastore/Registry writes in the opposite order would indeed present a challenging transition.

TallJimbo replied:
This is now better addressed in the text, as per our more recent Slack conversation.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+This change will make it so Datastore sees conflicting writes before Registry.
+While these are expected to be rare, and thus conflict resolution does not need to be highly optimized, is important that it does not have the potential to corrupt the repository.
+The easiest way to do this is to configure Datastore storage backends to prohibit writes to locations where files already exist; this is the default behavior for most backends of interest, but we currently explicitly permit silent clobbering to work around problems with automatic retries in BPS.
+:cite:`DMTN-205` outlines an alternative approach to automatric retries that is better for provenance and should be adopted instead.
```
Typo: automatric

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+2. Call ``Datastore.get`` with both the resolved ``DatasetRef`` and the bundle of records, returning the result to the caller.
+
+Because this is a read-only operation, consistency in the presence of failures is not a concern, but this still has a major advantage over the current approach for client-server in particular, as it bundles all http server access into a single call, followed by a direct object-store call, reducing latency and again allowing the Datastore to be a regular ``FileDatastore``.
```
This is all fine.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+``Butler.import`` and ``Butler.transfer_from``
+""""""""""""""""""""""""""""""""""""""""""""""
+
+These operations would behave like vectorized versions of ``Butler.put``, with all Datastore writes (if nontrivial transfers are required) occurring before a single Registry or butler server operation that (within a transaction) adds datasets and the associated datastore records to the database.
```
The main issue with transfer_from is the large numbers of records that are involved for 250k transfers in batch. On the plus side those are presumably not using client server and we will just have to have a records registry API that allows it to do what it does now inside Datastore which is to return 10k of records at a time to keep things manageable. We can defer all the dimension insertion and dataset insertion until after we have done the file transfers or do the dimensions up front still as you say below.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+
+These operations would behave like vectorized versions of ``Butler.put``, with all Datastore writes (if nontrivial transfers are required) occurring before a single Registry or butler server operation that (within a transaction) adds datasets and the associated datastore records to the database.
+
+For ``Butler.import``, however, we also need to add dimension data, and register collections and dataset types, and not all of these can be performed in transactions.
```
We do all of that in transfer_from as well.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+Adding journal files
+--------------------
+
+The main flaw in the proposal above is that it can leave artifacts in the Datastore root that are untracked and hard to find, due to both I/O failures and abandoned batch runs.
```
batch also has the problem of not writing an output file when we expected one. This is why transfer_from has to check for the existence of 250k files in an object store and is the main reason it's slow. We talked about having a per-quantum look up file which would describe the actual files that were written and that would speed things up.

TallJimbo replied:
I completely agree, and would still like to go in that direction someday.  But I think it's not a consistency problem so I wasn't going to get into it here.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+Bundling DatasetRef with Datastore records
+------------------------------------------
+
+- If we're passing around DatasetRefs and bundles of datastore records together a lot, we should have a class for that.
```
I'm guessing that @andy-slac has got very close to this in quantum graph. Since the system doesn't know anything until a specific datastore is created it has to be fairly generic. The "records" part of a `DatasetRef` would still have to be something like `dict[str, list[StoredDatastoreItemInfo]]` and also be a pydantic model for serialization purposes (the serialization would have to include the record subclass but I know how to do that).  `StoredDatastoreItemInfo` already has the UUID in it though so we have to be a bit careful -- there are cases where `list[StoredDatastoreItemInfo]` is fine (eg for `Butler.mexists()` calls) without having to include the hierarchy and memory bloat for `list[DatasetRef]` each carrying its own rows.

TallJimbo replied:
From the perspective of integration with the query system, there are some advantages to actually putting the datastore records into the `DatasetRef` and saying that an "expanded" `DatasetRef` has both expanded data IDs (with dimension records) and datastore records.  I was playing around with that in some prototyping on this ticket last week, and I'm not sure its an idea that holds up everywhere.  But I think it's at least intuitive from a user perspective (I do not expect users to go poking at the Datastore records in a `DatasetRef`, at least most of the time, but I think they may care that a query they do include them, so they can use those refs in other places without forcing more lookups).

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+------------------------------------------
+
+- I've totally come around to the idea of Butler being the only provider of public interfaces, and we should move all useful Registry methods there, at the same time we define a richer set of ABCs for various aspects of Butler interfaces.
+- Registry would become a nonpolymorphic concrete class used by one Butler subclass and the Butler server.
```
Are you sure? I was coming round to having the public API be `Butler.queryDatasets` (or in fact `Butler.query_datasets` if we are changing things anyhow) but internally that being written as `Butler._registry.query_datasets`. I don't really follow how we end up with a client/server registry at the Butler class level. I thought the purpose of the reorganization was to allow the transactions to be removed from Butler and have atomic Registry APIs that can be switched depending on direct database vs client/server.

TallJimbo replied:
I am by no means sure, especially after the discussion last Thursday.  The first thing I'm planning to prototype out for this section is now more along the lines of treating Registry as a private polymorphic implementation object, so that we'd have one concrete full Butler class with the full public interface, and different private Registry classes for SQL and client/server.

timj replied:
The Registry implementation class is already completely configurable from YAML. You can see that in the client/server registry implementation vs sql registry.

TallJimbo replied:
My time as an advocate of polymorphic butler was very brief.  Technote is fully back to polymorphic registry as the divide between SQL and client/servert.

timj commented on index.rst:
```diff
@@ -13,17 +13,256 @@
+- A lot of Registry's convenience functionality should move to Butler.
+- I think Registry's caching should move to Butler.
+- Registry's default collections and governor data ID values should definitely move to Butler.
+- Some of the convenience logic in the butler CLI scripts should also move to Butler.
```
I'm fine with this. It's design work but be careful because the API for the "script" layer is always simplified objects (strings, numbers etc) and not objects that people really want to be using.

TallJimbo replied:
I completely agree.  It's probably the display-results and confirmation-checking side of the scripts that need Python interfaces, not the "I'll run queryDatasets for you" bits.

timj reviewed (approved):
I am on board with many of these changes. I think that having an internal Datastore and internal Registry that the users never see will work and having them pass Datastore Records via the Butler is okay. These changes will make client/server butler much easier and I think will remove the need for a client server `Butler` itself.

I'm not sure I yet understand how we don't have a `butler._registry` pluggable registry class for client/server vs direct connection.

TallJimbo committed:
Add prototyping directory full of Python stubs.

This is far from complete, but I need to start committing it bits so
I can more easily change my mind as I go.

TallJimbo committed:
Remove duplicate copy of RemovalHelper and improve docs a bit.

TallJimbo committed:
Run isort and rerun black.

TallJimbo committed:
Work out deletion logic more thoroughly in prototyping.

TallJimbo committed:
Prototype get/put/etc implementations.

This also drops "exists" for now, since I don't know what that should
actually do.  DM-32940 is even trickier than I thought when you
consider LimitedButler as well.

It also drops "purge" more permanently - I think it does't make sense
for LimitedButler, which can't delete things from a Registry it
doesn't have, and full Butler has removal() as a safer and more
helpful way to do this.

TallJimbo committed:
Rename vectorized method variants.

Prefixing with 'm' was nice for simple names, but adding _many is
better more generally.

TallJimbo committed:
Add Query.join_external_uuids.

We'll need this to have Query support expanding DatasetRefs.

TallJimbo committed:
Fix import.

TallJimbo committed:
Work out batch-friendly edit operations.

import and transfer_from are the last big pieces, but I think they'll
mostly amount to making a RawBatch in different ways and then using
existing Registry and Datastore methods.

TallJimbo committed:
Work out Butler.import_ and start on export.

TallJimbo committed:
Add DatastoreButler ABC and rework opaque table data structures.

TallJimbo committed:
Fix dimension data batch data structures.

TallJimbo committed:
Finish sketching out transfers and import/export.

Docs for Datastore, LimitedButler, and DatastoreButler done.

TallJimbo committed:
Finish docstrings and README for prototyping directory.

TallJimbo committed:
Update text to be not-inconsistent with prototype.

This addresses the comments in the PR discusison about "staging areas"
(which we don't want) and tweaks some order-of-operations to reduce
back-and-forth (and hence high-latency) between Registry and Datastore.

timj commented on index.rst:
```diff
@@ -13,17 +13,249 @@
+
+.. note::
+   TODO: think about how all of this affects disassembled composites.
+   Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
```
We also talked about virtual composites. A dataset that could be assembled from components of multiple other datasets. This was how we thought we were going to create a "calibrated raw" by defining it to be a combination of the raw with the WCS, say, coming from the calexp, and the metadata coming from a recalculated header constructed from the EFD. This is also more efficient if the calexp WCS was stored as a separate component.

TallJimbo replied:
I think about the "virtual composites" use case now as more of a PipelineTask/QG problem.  Certainly for reconstructed (non-compressed) PVIs we'll need to run PipelineTask code that really shouldn't be invoked "under" the butler interfaces inside a Formatter, and while assembly of a "calibrated raw" might not rise to that threshold, it obviously _could_ be done be a PipelineTask, and I'd rather have one system for all of these kinds of things.

In both cases we do want to be able to read just the (lossless-compressed) mask and the structured-metadata objects like (at least) WCS from the lossy-compressed PVI, while getting the image plane and possibly a variance plane from something else. But in the `PipelineTask` approach that means the penalty for not disassembling is staging larger-than-necessary files to whatever server-side resources are used to run the task, rather than sending those files all the way to the user.  And I don't think our current disassembly represents a very good way to divide up files anyway:

 - Full disassembly of an Exposure along component lines is a _lot_ of really tiny files, and that's not great for overall I/O either (or inode usage), at least on filesystems.  If disassembly did not operate on a per-component basis, and we could define, say, 3-4 blobs to comprise a 15-component `Exposure`, with one blob aggregating a bunch of small things and the others being image/mask/variance, that'd be ideal for our expected I/O patterns (especially if it lets us gzip the small-things-together blob, which isn't going to be FITS-tile-compressable).  But if there are so few blobs, byte-range reads of a single file from the object starts to sound viable, too.

 - Our final DRP WCSs, PSFs, and backgrounds are all produced per-visit rather than per-{visit, detector}.  So our pipelines are also going to naturally produce visit-level FITS binary table datasets with all of the structured metadata that will go into a PVI, and one option for us is to just retain those for use in reconstructing PVIs and the like instead of reading them from the compressed PVIs.

Metadata coming from the EFD instead of processing is something I have not thought about, at least not in a long time.  Sounds like it'd have tricky provenance implications.

timj replied:
I don't want to abandon composite disassembly without wider discussion with at least @ktlim and @gpdf. Now that we have read-only components we could change how we do disassembly such that we split it into data, variance, mask + ExposureInfo.

For the virtual composites with the updated headers calculated from the EFD, the idea originally was that we would do a batch job where we regenerate the header components from the EFD and store them as distinct datasets in butler. Then on `butler.get` it would understand that the `raw+` dataset type should be constructed by reading the raw then reading the header and then replacing the header component before returning it to the user. This seemed entirely like something that butler itself should be able to manage rather than asking people to retrieve the raw and then go looking for the relevant header dataset. The `raw++` dataset that has the raw pixel data plus all the final PSF and WCS information attached was also wanted at one point but maybe that could still be done without disassembly, but by downloading both the raw and the compressed PVI and then having a virtual composite define which components comes from the raw and which comes from the PVI (and you have to read both into memory if they are not disassembled).

TallJimbo replied:
Ok, if the EFD information is getting there via a batch job that dumps the information into other butler datasets first, I'm much less concerned about it, as that batch job will have the kind of provenance I'm worried about.

And I don't think we need to be in a rush to drop composite assembly - I've been careful to leave room for datastore opaque tables that need more than just UUID (i.e. component) in their primary key, and I think that's the main potential concern here.

The DECam-raw case with multiple datasets in a single file is more worrisome, but mostly for deletes, I think, so if we declare that those can only be ingested via absolute (and hence "unowned") URIs, that may be fine, too.

ktlim replied:
It sounds like the suggestion is that any generation of "virtual" data products including combinations of components (but not extraction of single components) will require a Pipeline, either with the user adding "converter" PipelineTasks to an existing Pipeline or behind the scenes of an image retrieval/regeneration service.  I'm a little worried about potential overheads of doing things this way, but exposing the conversion via a PipelineTask rather than hiding it in the Butler is not inherently unacceptable.

timj replied:
> but mostly for deletes

Datastore currently has logic to handle the DECam case such that it only deletes the actual file if there are no more remaining records containing that file. This is easy when Datastore owns the records but sounds impossible if Registry is only telling Datastore what to delete without telling Datastore that other records exist for the file -- it seems that registry will have to have the logic of deciding for unstore whether to send any records to datastore at all, only sending records if all the records are relevant -- this then becomes problematic because then registry has to understand what a column in the records means. The whole point of Opaque is that registry doesn't know what the records mean in terms of real data files, or if indeed there are any files at all (which is another problem for signing -- the specific datastore is the only thing that knows that it has an actual file that can be converted to a public signed URL).

I imagine that in most cases "virtual" composites would be generated by pipelines somehow (or some batch system). The user cares about being able to retrieve something without knowing that it's virtual.

timj replied:
To be a little more concrete about virtual composites, I was thinking along the lines of a new table that would have:

```
UUID | UUID1 | Component|
abc    | def      |                     |
abc    | xyz      | wcs             |
abc    | mno     | header        |
```
and you read this as being a virtual dataset known to registry as abc, with the entire content of def modified by replacing component wcs from xyz and component header from mno. Maybe we no longer need that sort of functionality. The StorageClass delegates we already have written and the component handling code in datastore make this fairly straightforward for reading.

TallJimbo replied:
I was hoping we could implement that sort of functionality via recipes that operate on dataset types and data IDs, and thus don't need any storage and could maybe even be implemented outside butler (though I also suspect users want it available via `butler.get`).

TallJimbo committed:
Update to reflect new approach to where signed URLs are made.

ktlim commented on index.rst:
```diff
@@ -216,6 +213,34 @@ Which of these is preferable probably depends on whether we want these operation
+- It requires the Datastore to have a server component; in every other respect we can abstract the differences between a SQL-backed full Butler and a client/server full Butler via a different Registry implementation (see :ref:`public-interface-changes`).
+
+- Because the information used to determine whether a URL *should* be signed lives in the Registry, a Datastore server cannot easily perform this job.
+  It cannot trust information provided to it via the `DatasetRef` and opaque table records passed to it from Butler in terms of access control, since those come from a public http interface; instead it'd have to use a Registry of its own to re-obtain Datastore records and collection information for each UUID it is passed in order to determine whether to sign URLs embedded in the records.
```
Unless the Registry provides a signature over the access control information (at least) that can be passed through but not independently generated by the external client.

TallJimbo replied:
I'm not sure I grokked "signature over the access control information" - you mean something cryptographically derived from the access control information and the UUID that the server could verify without access to the registry?

timj replied:
In the other tech note the datastore had an internal registry so the datastore server handling all this was easy. Now that we want registry server to report the records the registry server has to return records that include signed urls. Is there a problem with that? The main subtlety is that signing for read and signing for write are different things so the server needs to do something different when put or ingest is involved.

TallJimbo replied:
> Now that we want registry server to report the records the registry server has to return records that include signed urls.

That's exactly what this most recent commit does.  Before this commit the datastore would have had to have that internal registry; I just didn't realize it.

ktlim replied:
"signature over the access control information" means something like hashing the critical information along with a shared secret between the registry and the datastore servers.  That allows information to be passed via the client without the client being able to change it.

ktlim replied:
I still think this paragraph is wrong as written.  Datastore does not have to re-obtain Datastore records.  It only needs to share a secret with the Registry and verify a signature over the records.  This is similar to signing URLs, but it's signing records instead.  Issues of lifetime and replay are the same.

TallJimbo replied:
I've [updated](3d54cdc01fec000a455acb2c4430376861e8cd49) the text to reflect this possibility.  I haven't changed my overall recommendation because the original proposal still wins on reducing the number of server round-trips and complexity (though my complexity metric gives a lot more weight to http client/server interfaces than regular Python interfaces, and that may be a lack-of-familiarity bias).

TallJimbo committed:
Tell sphinx to ignore prototyping/README.rst

TallJimbo committed:
Provide introduction to public API changes in prototype.

It's clear in this that this document is really missing a way to link
effectively from the prose to the prototype code, but I'm not sure
that's actually worth any time investment.

TallJimbo committed:
Unify and expand discussion about Datastore races and clobbering.

TallJimbo committed:
Finish first draft.

This adds an introduction that's mostly about motivation and more
caveats on API changes, replaces TODO notes with what I think are the
most useful statements we can make right now, and dials down the
ambition for this ticket w.r.t. QuantumGraph execution design work.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+   This state is expected to be transitory, either intentionally (e.g. during batch execution, before datasets are transferred back), or as a result of failures we cannot rigorously prevent.
+   Datasets in this state as a result of failures or abandoned batch runs are considered undesirable but tolerable, and an approach to minimizing them will be introduced later in :ref:`adding_journal_files`.
+
+2. When a dataset is present in both Registry and Datastore, the Registry is fully responsible for storage of Datastore records.
```
Can we have an early definition/example of "Datastore records"?  Are these one per dataset?  Sounds like one per file (== component)?  Schema is variable, but must include UUID — how variable?

TallJimbo replied:
Done; short answer is that they need to have the UUID _in_ their primary key, but a compound primary key that also has other fields is permitted.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+   Datasets in this state as a result of failures or abandoned batch runs are considered undesirable but tolerable, and an approach to minimizing them will be introduced later in :ref:`adding_journal_files`.
+
+2. When a dataset is present in both Registry and Datastore, the Registry is fully responsible for storage of Datastore records.
+   It will be the job of the Butler to accept records from Datastore and pass them to Registry on ``put``, and to fetch them from Registry and pass them to Datastore on ``get``.
```
Why does Datastore not talk to Registry itself but must always be intermediated through Butler?

TallJimbo replied:
I've removed this sentence in favor of one that is less prescriptive, as the original statement is actually now directly contradicted later in the text.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
+
+- for a client/server butler, there is little alternation between object-store and http operations, reducing latency (assuming the data ID has indeed been obtained in advance) and increasing the possibility that the client Datastore can just be a regular ``FileDatastore``.
```
"little alternation between object-store and http operations"?  This means bundling operations into a single server call?

TallJimbo replied:
Yes, exactly. I've rewritten slightly to say "bundle operations" directly.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+   A POSIX-backed Datastore could handle these races by writing to a temporary random location and hard-linking into the final location, since hard-links are atomic and do not clobber by default.
+   Object-store Datastores do not have this option, and do not in general provide the kind of one-writer-succeeds behavior we'd need.
+   This is largely mitigated by the fact that QuantumGraph execution provides high-level management of possible races (as long as there is only one QuantumGraph for any RUN collection).
+   It could be further mitigated by including the UUID in the URI template, making it far less likely that the Datastore writes will clash; when the Registry transactions do clash, at most one will be committed (and associated with a valid Datastore write), while the others will roll back, leaving some Datastore-only datasets (which, again, are permitted albeit undesirable).
```
Under what circumstances does the same dataset have different UUIDs?

Doesn't adding the UUID to the URI template make it more difficult/impossible for non-Butler utilities to work with dataset files/objects?  Should this be a consideration?

TallJimbo replied:
The only one I can think of now is the classic naive-user mistake of using external parallelization to run `pipetask` (e.g. direct slurm submissions) without making a single QG up front.  So I think this isn't much of a mitigation at all, and I'll just drop the paragraph.

That does make it clear that this division of responsibilities leaves us depending quite a bit on higher-level tools to prevent data races, and that makes me a bit uncomfortable.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+This is a proposed new interface for removing multiple datasets from the Datastore without removing them from the Registry - one part of a replacement for ``Butler.pruneDatasets``, and part of a reimplementation for ``Butler.removeRuns``.
+
+1. Pass the inputs to Registry and/or butler server to obtain ``DatasetRefs`` and datastore records, instructing it to delete those records at the same time.
+   ``QuantumBackedButler`` may not need to implement this operation at all, but if it does (e.g. for clobbering), it already has everything it needs in the quantum.
```
Can QuantumBackedButler delete datastore records?

TallJimbo replied:
It certainly can't remove them persistently from the quantum it read itself, as it is not allowed to write to the QuantumGraph file where that was stored.  I think whether it deletes anything in memory is an implementation detail.  And while it can write a summary of what it did for downstream consumers (like the transfer job) that could include an indication that a predicted dataset was not written, I think that means the whole `DatasetRef`, not just particular records.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+A SQL-backed Datastore that transforms in-memory datasets fully into Datastore records would not need to use journal files at all, and a purely ephemeral in-memory Datastore could use in-memory objects to store journal content instead of files.
+
+One unique and particularly important type of journal file is one that signals an ongoing QuantumGraph execution that has not yet been transferred back.
+This could be a pointer to the QuantumGraph file or even the QuantumGraph file itself, since a QuantumGraph already carries all the information needed to find all datasets that may have been written and not transferred back as part of its execution.
```
QG journal file is assuming that datasets are `put()` to Datastore as they are generated, not in a batch at the end?

TallJimbo replied:
That's where they're most useful as journal files, certainly.  If the files are written to scratch space and actually transferred later, it wouldn't be _incorrect_ to consider the QG a journal file (since a journal file is permitted to exist even if none of the datasets it refers to are present in the Datastore), but a journal file with a shorter, transfer-only duration would be more efficient.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+
+``QuantumBackedButler`` will not write or delete journal files; it will rely entirely on a higher-level one for the full QuantumGraph.
+
+For a client/server Butler the journal creation and deletion could happen in the client or in the server, but the former continues to permit a regular ``FileDatastore`` to be used.
```
If the journal is client-side, it could be lost while still having Datastore-only files on the server.

If the journal is server-side, we need to protect against people spewing journals or reading others'.

If journals are server-side, could they be placed in the Registry database rather than in files?

How do we protect against journal files including other people's URIs?

Do journal files include signed URIs?  If so, could there be a problem cleaning up much later after they expire, or do we need to refresh them?

TallJimbo replied:
I was thinking of journal files as living somewhere under the datastore root, and I completely agree that journal files on some client-only storage are not viable; by client-side I meant "written directly by a Datastore client to the remote object store".

I hadn't previously thought about the other issues you've raised here, and I've now [updated](c1805e2856e43d721b928895e952c50d37bad8ea) the text to address them:
 - I've deleted this paragraph and the longer one about journal file locations in client/server under the `unstore` operation.
 - I've added a new subsection at the end of the "Including signed URLs" section to discuss the interplay between journal files and access control.

I've also updated the prototype Python code to have the server provide signed journal-file URLs for a client-side Datastore to use, as I think that's the simplest approach, but one in which there are server calls for making journal files is also viable and that's reflected in the text.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+
+This constitutes a new definition of an "expanded" `DatasetRef`: one that holds not just an expanded data ID, but a bundle of datastore records as well.
+
+Combined with the conclusion of :ref:`including-signed-urls-for-access-control`, this means we'd be returning signed URLs in the `DatasetRef` objects returned by query methods.
```
Including URIs in DatasetRefs may encourage people to directly access them, which would prevent auto-refresh (which requires a server round-trip to verify the DatasetRef contents).  Need to ensure that only approved interfaces are used.

TallJimbo replied:
I had been mostly ambivalent about the extent to which the Datastor records would be private or public attributes of a `DatasetRef`, and I *think* this is "just" a (strong) argument for making them private.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+----------------------------------------
+
+Our access control model for the official Rubin Observatory data repositories (see :cite:`DMTN-169`) is based on information stored in the Registry - collection names, whenever possible, and new naming conventions or new columns in contexts (e.g. dimension records or dataset type names) that are not associated with collections.
+Access to the associated files managed by a Datastore is mediated by signed URLs; once the server side of the remote Butler has determined that an API call is permitted (based on that the Registry-side information), it can generate one or more signed URLs to pass to Datastore that provide direct access to the controlled files.
```
Will signed URLs go through Gafaelfawr and be able to be rate-limited?

TallJimbo replied:
If we need to go through Gafaelfawr to sign URLs, I think that's ok, as long as we can do it in bulk rather than one at a time.  But I think the Butler server will have to contact Gafaelfawr up front to authenticate each request and obtain some sort of summary of what that user can access, and from that the Butler server should have everything it needs to know _whether_ to sign a URL (or even perform a query against a particular collection) without contacting Gafaelfawr again.

Rate-limiting between Butler server and whatever service it calls to do the signing sounds like it might make having an async server implementation even more important, but otherwise I don't see an obvious problem with it.

I'm pretty out of my depth on this part, though, so feel free to instead tell me how you think it ought to be done and I can tell you if that's problematic from my end :slightly_smiling_face: .

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+This only works rigorously for Registry operations at present, as the Datastore transaction system we have at present is not fault tolerant.
+Much of the rest of this proposal is motivated by trying to address this, but our current transaction interface is not really viable even for Registry-only operations when an http client/server implementation is required.
+
+Instead, the prototype includes a `RawBatch` class that represents a low-level serializable batch of multiple `Registry` operations to be performed together within one transaction, and a `Butler.batched` method and `BatchHelper` class to provide a high-level interface for constructing them.
```
Having client-side code prepare RawBatches could lead to extended locking of the database.  Can we mitigate by somehow limiting the size of RawBatches?

TallJimbo replied:
Shouldn't be hard at all - we can pretty easily get a rough estimate of the number of rows modified by a raw batch, and just put a cap on that in both the client (for convenience) and server (for rigor).

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+Combined with the conclusion of :ref:`including-signed-urls-for-access-control`, this means we'd be returning signed URLs in the `DatasetRef` objects returned by query methods.
+This is mostly a good thing - it makes those refs usable for reading datasets directly and it completely avoids redundant registry lookups in usual workflows.
+But it does increase the duration we'd want a typical signed URI to be valid for - instead of the time it takes to do a single operation, it'd be more like the time it takes the results of a query in one cell of a notebook to be used in another cell.
+While that's arbitrarily long in general, I don't think it's unreasonable to either tell users that refs will get stale after a while, or just add timestamps to signed URIs so we can spot them and refresh them as necessary when ``get`` is called.
```
Object-store signed URIs already have timestamps in them.

TallJimbo replied:
Are those timestamps hashed in somehow, or are they something client could directly read to decide whether it needs to make server call to refresh them at all?

ktlim replied:
In practice, the timestamp is visible (as an `Expires` parameter in the URL), but in theory the URL format could change at any time.

ktlim commented on index.rst:
```diff
@@ -10,20 +10,357 @@
+A seemingly natural place to include URL-signing in this proposal is inside the Datastore implementation, since only the Datastore knows where any URLs might exist in the opaque-to-Registry records it uses, and only the Datastore ever uses and kind of URLs.
+This approach has two major drawbacks, however:
+
+- It requires the Datastore to have a server component; in every other respect we can abstract the differences between a SQL-backed full Butler and a client/server full Butler via a different Registry implementation (see :ref:`public-interface-changes`).
```
If Datastore is only client-side but is providing records for server-side Registry to store, we need to make sure they're for proper UUIDs (accessible to the user) and we should protect against excessive resource consumption.

TallJimbo replied:
I didn't understand the concerns here - proper UUIDs in the sense that they really can't clash?  Accessible to the user in what sense?  Are you worried about clients sending huge batches of records to the server and eating up server resources as a result?

In any case, I've reworded this to make it clear that there will be server-side Datastore logic regardless, and that includes things like schema definition and at least the initial versions of all of the Datastore records - the client-side modifications are things like file sizes and checksums, so if we need to have the server check that Datastore records coming from the client are reasonable, it should have everything it needs to do that.

TallJimbo committed:
Handle access control in journal files more carefully.

TallJimbo committed:
Explain "Datastore records" up front.

TallJimbo committed:
Remove outdated statement about Butler doing all mediation.

TallJimbo committed:
Remove UUIDs-in-templates as mitigation for data races.

On further inspection that wouldn't actually help with the problem.

TallJimbo committed:
Include server-side shared secrets as an alternative way to sign URLs.

TallJimbo committed:
Minor text edits and clarifications.

TallJimbo committed:
fixup! Handle access control in journal files more carefully.
